### PR TITLE
52 login esnaccounts

### LINKED ISSUE #52: Login with ESNAccounts API
> As participant, I'd like to login with my ESN account so that I don't have to keep track of which emails I am using, rendering the experience easier.

- [ ] Add the login with ESN accounts based off the current GA app
- [ ] Keep the current login with Cognito as an alternative for externals
- [ ] In the  login page add the options to login as an ESNer or login as an external.

### PR DESCRIPTION
**Pull request by rbento1096**
closes #52

- [ ] currently can't use cognito on swagger for authorization
- [ ] we substitued 'configurations' in ESN-GA app for 'Roles' here. Does it make sense?
- [ ] For logging out, can we use the same, method for ESN and cognito?

@uatisdeproblem can you give a check to the code to make sure everything is okay and also the points above? Not super urgent, just to make sure everything is ok

### CODE REVIEW

**Comment by uatisdeproblem** [comment]
Ok, prossima settimana guardo con calma

**Review by uatisdeproblem — COMMENTED** [review]
_EDIT: translated in english._

Sorry if I'm reviewing this only now, but I didn't have much time over the weekend.

I've reviewed everything except the `auth.guard.ts` file because I want to take a moment to figure out how to make the Cognito (guest) login and the ESN Accounts login work together.

I'll look into it this week, in the meantime, I'll leave you the rest of the review if you want to look at those points.

File: back-end/src/handlers/login.ts [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+      );
+
+      const user = new User({
+        userId,
+        email: attributes['cas:mail'][0],
+        sectionCode: attributes['cas:sc'][0],
+        firstName: attributes['cas:first'][0],
+        lastName: attributes['cas:last'][0],
+        roles: attributes['cas:roles'],
+        section: attributes['cas:section'][0],
```
[the remark this answers is not in the document]
Comment (rbento1096) [reply back-end/src/handlers/login.ts:76]: Is it fixed? Right now I got all users with correct sections and countries on the dyanmo db data

File: back-end/src/models/roles.model.ts [hunk 2]
```diff
@@ -0,0 +1,30 @@
+import { Resource } from 'idea-toolbox';
+
+/**
+ * The platform's roles.
+ */
+export class UserRoles extends Resource {
+  /**
+   * A fixed string, to identify the role.
+   */
+  PK: string;
```
[the remark this answers is not in the document]
Comment (rbento1096) [reply back-end/src/models/roles.model.ts:10]: ah ignore this, it has been eliinated on the successive commit (un published) once I had created the user profiles

**Comment by rbento1096** [comment]
> Yes, but based on the type of login you have to make two different actions; therefore, I would add the type of user in the model's attributes

Need to do this after fixing the login with Cognito

> What kind of "roles" or "permissions" did you think of? Except for "administrator" and "standard user". So I can give you a more reasoned opinion.

As i said above I abandoned this logic. It is now an ENUM as is standard in IDEA. The full ENUM is here: 

```typescript
export enum Roles {
  NONE = 0, // starting user
  PARTICIPANT = 10, // user with an assigned spot
  DELEGATION_LEADER = 20, // same as participant, but can see delegation (in future)
  STAFF = 30, // will have some more permissions during the physical event
  OC = 40,  // these 3 might all be mixed into one, honestly, I added as a safekeep
  CT = 50,
  IB = 60,
  ADMIN = 70 // max permissions
}
```

Currently I'm only checking for ADMIN on certain actions because I still need Adel to confirm me which roles should do what.

File: back-end/src/handlers/auth.ts [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+
+///
+/// CONSTANTS, ENVIRONMENT VARIABLES, HANDLER
+///
+
+const DDB_TABLES = { roles: process.env.DDB_TABLE_roles };
+const ddb = new DynamoDB();
+
+const SECRETS_PATH = 'esn-ga/auth';
+const secretsManager = new SecretsManager();
```
[the remark this answers is not in the document]
Comment (rbento1096) [reply back-end/src/handlers/auth.ts:20]: done!

File: back-end/src/handlers/login.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (uatisdeproblem) [reply back-end/src/handlers/login.ts:76]: Confirmed: it was solved somewhere in the last few months: now it also works in the GA app

**Pull request merged**